src/coreclr/jit/loopcloning.cpp

```diff
@@ -2101,40 +2023,50 @@ void Compiler::optCloneLoop(unsigned loopInd, LoopCloneContext* context)
     // to this pair of loops:
     //
     // H ?-> H3   (all loop failure conditions branch to new slow path loop head)
-    // H2--> E    (Optional; if T == E, let H fall through to T/E)
+    // H2 --> E   (if T == E, H2 will fall through to T/E)
     // T
     // E
     // B  ?-> T
-    // X2--> X
+    // X2 --> X   (only need this if B -> is conditional, not BBJ_ALWAYS)
     // H3 --> E2  (aka slowHead. Or, H3 falls through to T2 == E2)
     // T2
     // E2
     // B2 ?-> T2
     // X
 
     BasicBlock* h = loop.lpHead;
-    if (!h->KindIs(BBJ_NONE, BBJ_ALWAYS))
-    {
-        // Make a new block to be the unique entry to the loop.
-        JITDUMP("Create new unique single-successor entry to loop\n");
-        assert((h->bbJumpKind == BBJ_COND) && (h->bbNext == loop.lpEntry));
-        BasicBlock* newH = fgNewBBafter(BBJ_NONE, h, /*extendRegion*/ true);
-        JITDUMP("Adding " FMT_BB " after " FMT_BB "\n", newH->bbNum, h->bbNum);
-        newH->bbWeight = newH->isRunRarely() ? BB_ZERO_WEIGHT : ambientWeight;
-        BlockSetOps::Assign(this, newH->bbReach, h->bbReach);
-        // This is in the scope of a surrounding loop, if one exists -- the parent of the loop we're cloning.
-        newH->bbNatLoopNum = ambientLoop;
-        optUpdateLoopHead(loopInd, h, newH);
+    assert((h->bbFlags & BBF_LOOP_PREHEADER) != 0);
 
-        fgAddRefPred(newH, h); // Add h->newH pred edge
-        JITDUMP("Adding " FMT_BB " -> " FMT_BB "\n", h->bbNum, newH->bbNum);
-        fgReplacePred(newH->bbNext, h, newH); // Replace pred in COND fall-through block.
-        JITDUMP("Replace " FMT_BB " -> " FMT_BB " with " FMT_BB " -> " FMT_BB "\n", h->bbNum, newH->bbNext->bbNum,
-                newH->bbNum, newH->bbNext->bbNum);
+    // Make a new pre-header block 'h2' for the loop. 'h' will fall through to 'h2'.
+    JITDUMP("Create new header block for loop\n");
 
-        h = newH;
+    BasicBlock* h2 = fgNewBBafter(BBJ_NONE, h, /*extendRegion*/ true);
+    JITDUMP("Adding " FMT_BB " after " FMT_BB "\n", h2->bbNum, h->bbNum);
+    h2->bbWeight     = h2->isRunRarely() ? BB_ZERO_WEIGHT : ambientWeight;
+    h2->bbNatLoopNum = ambientLoop;
+    h2->bbFlags |= BBF_LOOP_PREHEADER;
+
+    if (h->bbJumpKind != BBJ_NONE)
+    {
+        assert(h->bbJumpKind == BBJ_ALWAYS);
+        assert(h->bbJumpDest == loop.lpEntry);
+        h2->bbJumpKind = BBJ_ALWAYS;
+        h2->bbJumpDest = loop.lpEntry;
     }
-    assert(h == loop.lpHead);
+
+    fgReplacePred(loop.lpEntry, h, h2);
+    JITDUMP("Replace " FMT_BB " -> " FMT_BB " with " FMT_BB " -> " FMT_BB "\n", h->bbNum, loop.lpEntry->bbNum,
+            h2->bbNum, loop.lpEntry->bbNum);
+
+    // 'h' is no longer the loop head; 'h2' is!
```

Comment:
Have this code after the line `h2->bbFlags |= BBF_LOOP_PREHEADER;` ?

```c++
    // 'h' is no longer the loop head; 'h2' is!
    h->bbFlags &= ~BBF_LOOP_PREHEADER;
    optUpdateLoopHead(loopInd, h, h2);
```

Reply:
There are several orders that would work. IMO, I like this here because I want "all" the updates to `h2` to happen first.